### PR TITLE
[RFC] TensorFlow Lite Micro as a module

### PR DESCRIPTION
**Pull request by laurenmurphyx64**
## Issue
#34295 

## Introduction
Adds the TensorFlow Lite Micro library as a Zephyr module and incorporates two TensorFlow Lite Micro samples into the tree for reference.

## Problem description
Unless they're willing to do legwork similar to this PR on their own, developers of TensorFlow apps on Zephyr build inside the TensorFlow tree with a Makefile generated from a template that invokes CMake and Make. Developers are not able to interact directly with `CMakeLists.txt`, `prj.conf` and other build settings / options in the ways to which they may be accustomed.

## Proposed change

Allow users to build TensorFlow apps on Zephyr in the usual way (i.e. with west).

- Add TensorFlow as a Zephyr module with a Kconfig option
- Incorporate two TensorFlow Lite Micro samples for reference

### Proposed change (detailed)

- Add TensorFlow 2.4.1 with static copies of third-party library dependencies
  - (See [the RFC-version of the TensorFlow Zephyr module repository](https://github.com/laurenmurphyx64/tensorflow). This is temporary and will be updated with a reference to the official Zephyr Project's TensorFlow repository.)
  -  Avoid replicating bash scripts to pull and patch these copies
  - (See `Dependencies` and `Concerns and Unresolved Questions` for further discussion)
- Add a single Kconfig option for enabling TensorFlow (for now) (`CONFIG_TENSORFLOW`)
- Add TensorFlow Hello World sample for most boards (build and run test case for QEMU in CI)
- Add TensorFlow Magic Wand sample for LiteX/VexRiscV emulated by Renode (build-only test case in CI)

### Dependencies
- [TensorFlow](https://github.com/tensorflow/tensorflow) [2.4.1](https://github.com/tensorflow/tensorflow/releases/tag/v2.4.1) ([Apache 2.0 License](https://github.com/tensorflow/tensorflow/blob/master/LICENSE))
- [Antmicro's LiteX/VexRiscv demo with emulation on Renode](https://github.com/antmicro/litex-vexriscv-tensorflow-lite-demo/) ([Apache-2.0 License](https://github.com/antmicro/litex-vexriscv-tensorflow-lite-demo/blob/master/LICENSE)) 
  - See also [their blog post](https://antmicro.com/blog/2019/12/tflite-in-zephyr-on-litex-vexriscv/), which was also [cross-posted onto the Zephyr blog in 2019](https://www.zephyrproject.org/tensorflow-lite-in-zephyr-rtos-on-litex-vexriscv/) with an [update in 2020](https://www.zephyrproject.org/running-and-testing-tensorflow-lite-on-microcontrollers-without-hardware-in-renode/)
- [FlatBuffers](https://github.com/google/flatbuffers) by Google ([Apache 2.0 License](https://github.com/google/flatbuffers/blob/master/LICENSE.txt))
- [Gemmlowp](https://github.com/google/gemmlowp) by Google ([Apache 2.0 License](https://github.com/google/gemmlowp/blob/master/LICENSE))
- [Ruy](https://github.com/google/ruy) by Google ([Apache 2.0 License](https://github.com/google/ruy/blob/master/LICENSE))

#### Note

In the TensorFlow build system, the last three are pulled in as specific versions from mirror repositories with only certain files and patched by TensorFlow's bash scripts. I copied these patched files into a `third_party_static/` static directory in [my RFC version](https://github.com/laurenmurphyx64/tensorflow/) of the TensorFlow Zephyr module repository. See [the commit history of that repository](https://github.com/laurenmurphyx64/tensorflow/commits/zephyr) for more information about these dependencies.

### Future porting of the Magic Wand sample to an actual board
I've been working on porting the Magic Wand sample to `frdm_k64f` for use with the FXOS8700 accelerometer, but hit a roadblock. The port will be added in a future PR.

### Building for LiteX VexRiscV in CI
If the GitHub CI doesn't already build for `litex_vexriscv`, we would need to add that so the Magic Wand sample is built along with the other samples.

### Static Third-Party Libraries

As mentioned before, TensorFlow uses several bash scripts to generate a Makefile from a template and download and patch certain files of specific versions of three third-party dependencies. Instead of attempting to duplicate that, I put a static copy of these files into my RFC version of the TensorFlow repository. We might want to think about whether it's a good idea (and if it's possible) to replicate those scripts and invoke them in the Zephyr build process. 

Regardless of what we decide to do, the owner of the TensorFlow Zephyr module will need to monitor the upstream to make sure we get any updates on the dependencies. The versions as well as the patches applied by the bash scripts might change in future releases of TensorFlow.

- [FlatBuffers commit](https://github.com/google/flatbuffers/commit/dca12522a9f9e37f126ab925fd385c807ab4f84e) dated September 24, 2020
- [Gemmlowp commit](https://github.com/google/gemmlowp/commit/719139ce755a0f31cbf1c37f7f98adcc7fc9f425) dated September 4, 2018
- [Ruy commit](https://github.com/google/ruy/commit/4790797d11a81f96baf24f3731fd3ca44c2c5f8b) dated November 15, 2020
- [TensorFlow download_and_extract.sh](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/tools/make/download_and_extract.sh) last updated January 5th

#### Magic Wand

```
ERROR: Test checkpatch failed: -:1166: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV) 
#1166: FILE: samples/tensorflow/magic_wand/src/accelerometer_handler.h:26: 
+extern TfLiteStatus SetupAccelerometer(tflite::ErrorReporter *error_reporter); 
                                                              ^ 
ERROR: SPACING: need consistent spacing around '*' (ctx:WxV) 
#1167: FILE: samples/tensorflow/magic_wand/src/accelerometer_handler.h:27: 
+extern bool ReadAccelerometer(tflite::ErrorReporter *error_reporter, 
                                                     ^
ERROR: SPACING: need consistent spacing around '*' (ctx:WxV) 
#3354: FILE: samples/tensorflow/magic_wand/src/output_handler.h:23: 
+void HandleOutput(tflite::ErrorReporter *error_reporter, int kind); 
                                         ^ 
```
As best as I can tell, this is [a bug in checkpatch](https://www.google.com/search?q=ERROR%3A+SPACING%3A+need+consistent+spacing+around+%27*%27&oq=ERROR%3A+SPACING%3A+need+consistent+spacing+around+%27*%27&aqs=chrome..69i57j69i58.165j0j7&client=ubuntu&sourceid=chrome&ie=UTF-8) where it mistakenly detects * as a multiplication symbol.

```
ERROR: GLOBAL_INITIALISERS: do not initialise globals to 0 
#1230: FILE: samples/tensorflow/magic_wand/src/constants.h:25: 
+constexpr int kWingGesture = 0; 
```
I think this might be another C/C++ issue with checkpatch. If I don't initialize this variable, the app won't compile.

#### Hello World

```
ERROR: Test checkpatch failed: Complex regular subexpression recursion limit (32766) exceeded at ~/zephyr-external/zephyr/scripts/checkpatch.pl line 2890.
Complex regular subexpression recursion limit (32766) exceeded at ~/zephyr-external/zephyr/scripts/checkpatch.pl line 2892.
Complex regular subexpression recursion limit (32766) exceeded at ~/zephyr-external/zephyr/scripts/checkpatch.pl line 2890.
Complex regular subexpression recursion limit (32766) exceeded at ~/zephyr-external/zephyr/scripts/checkpatch.pl line 2892.
Complex regular subexpression recursion limit (32766) exceeded at ~/zephyr-external/zephyr/scripts/checkpatch.pl line 2890.
Complex regular subexpression recursion limit (32766) exceeded at ~/zephyr-external/zephyr/scripts/checkpatch.pl line 2892.
```
This seems like more of a checkpatch issue, but I'm looking into it. If y'all have any thoughts on what might be causing this, please let me know.

```
-:857: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
#857: FILE: samples/tensorflow/hello_world/src/output_handler.h:24:
+void HandleOutput(tflite::ErrorReporter *error_reporter, float x_value,
```
Same bug.

### Python Lint Warnings

To avoid changing the machine-learning related code I'm not familiar with, I added flags to disable warnings for some pylint violations in the Python files under `samples/tensorflow/magic_wand/train/`. I can rewrite the code to remove all pylint violations if y'all prefer. 

## Alternatives
N/A

### CODE REVIEW

**Comment by laurenmurphyx64** [comment]
I noticed that the Renode repository had some updates in the last month after not being updated since November and the file with the MIT license has been removed. I will look more into the possibility of removing this file from our copy.

**Comment by laurenmurphyx64** [comment]
Updates:
- Added a license to the file `samples/tensorflow/magic_wand/renode/litex-vexriscv-tflite.repl`
- Removed MIT-licensed file `samples/tensorflow/magic_wand/renode/LiteX_I2C_Zephyr.cs` by upgrading Renode to 1.12

**Comment by nashif** [comment]
resolve conflicts please

**Comment by PiotrZierhoffer** [comment]
@laurenmurphyx64 for the record - the file with the MIT license is now part of Renode.

Right now we are working with this repo https://github.com/antmicro/tensorflow-zephyr-vexriscv-examples - this is similar to the one you mentioned, but tests current TFLite during nightly builds. You can read more about this CI setups on our [blog](https://antmicro.com/blog/2021/03/renode-based-ci-for-tensorflow-lite-micro/).

Of course it uses the "Tensorflow-centric" build procedure.

I will look more into this RFC soon, we are very interested in this.

**Comment by hakehuang** [comment]
if we can not modify the tensflow micro repo to enable CMake support how could we suppose make it as a mudole? I have created a PR to tensflow for this purpose, but was rejected due to they do not want to add build system support.
https://github.com/tensorflow/tensorflow/pull/47240

**Comment by carlescufi** [comment]
@laurenmurphyx64 thank you for this contribution! would you mind following the process described in [this page](https://docs.zephyrproject.org/latest/contribute/external.html) to fill out a new issue from the provided template with the information required (most of which is already in the description in the PR).

**Comment by laurenmurphyx64** [comment]
Hi @hakehuang! Thank you for your work so far on integrating TensorFlow with Zephyr and CMake. Yes, since we can't modify their repository directly, the Zephyr Project GitHub will host a modified copy of the TensorFlow repository following [the guidelines for Zephyr modules](https://docs.zephyrproject.org/latest/guides/modules.html#modules). There will be a maintainer who makes sure that it updates with each release upstream. You can see the [RFC-version of this copy on my GitHub](https://github.com/laurenmurphyx64/tensorflow) and see the modifications I made on [this commit](https://github.com/laurenmurphyx64/tensorflow/commit/203ede03c06a27de671fc85c7bda9ee2806cd8c0).

File: samples/tensorflow/magic_wand/train/requirements.txt
Comment (hakehuang) [inline samples/tensorflow/magic_wand/train/requirements.txt:2]: 2.4.1? or latest?

File: west.yml
Comment (hakehuang) [inline west.yml:25]: should this move to zephyr hosted repos?

**Review by hakehuang — COMMENTED** [review]
the C source code are ended with .cc but actually no c++ related things are called, so shall we change them to .c?

File: samples/tensorflow/magic_wand/src/main.cc
Comment (hakehuang) [inline samples/tensorflow/magic_wand/src/main.cc:27]: there is no zephyr rtos related things called, can we move the loop into a thread?

Comment (nashif) [reply samples/tensorflow/magic_wand/src/main.cc:27]: this is running in the main thread, so that should be fine.

File: modules/Kconfig.tensorflow
Comment (laurenmurphyx64) [reply modules/Kconfig.tensorflow]: Renamed to TENSORFLOW_LITE_MICRO :+1:

File: samples/tensorflow/magic_wand/train/requirements.txt
Comment (laurenmurphyx64) [reply samples/tensorflow/magic_wand/train/requirements.txt]: I checked the upstream and on 2.4.1, it is 2.4.1... so there's a chance that the copy I adapted on this commit doesn't match the original on 2.4.1 upstream. Thanks for pointing this out. I'll be checking this over soon.

File: west.yml
Comment (laurenmurphyx64) [reply west.yml:25]: We will move it after this PR is reviewed and before it's merged :+1:

**Comment by laurenmurphyx64** [comment]
> the C source code are ended with .cc but actually no c++ related things are called, so shall we change them to .c?

Some of them have some C++ syntax, particularly the Magic Wand files. I was able to change a couple files from .cc extension to .c extension in the Hello World sample, though.

File: samples/tensorflow/magic_wand/train/requirements.txt
Comment (laurenmurphyx64) [reply samples/tensorflow/magic_wand/train/requirements.txt]: So, after checking upstream and checking out the 2.4.1 release, I found out I was wrong and the tensorflow package version number upstream for the 2.4.1 release is 2.0.0-beta. Since I have some problems running the converted model following the `train/README.md` after upgrading the `requirements.txt` for tensorflow 2.4.0 (and numpy 1.19.2), I think it might be best to leave those alone for the time being:

image

Do you have thoughts?

Details: [The file in the 2.4.1 release](https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/lite/micro/examples/magic_wand/train/requirements.txt) has "2.0.0-beta1" in it. My confusion came from [the file's history on the main branch](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/magic_wand/train/requirements.txt) changing the version number to 2.4.0 being backdated to December even though the release came out in January.

**Review by carlescufi — CHANGES_REQUESTED** [review]
The CMakeLists.txt should be in the main tree, in the `modules/tensorflow` folder. That is the direction we are taking now, in order to maintain the module repo as free of zephyr-specific code as possible. CC @tejlmand

**Comment by laurenmurphyx64** [comment]
I've moved the CMakeLists.txt into the modules/tensorflow directory and changed the module's zephyr/module.yml following nanopb's example. I also split up the commit for each sample into two commits such that the first commit has the sample and the second has the Python training scripts.

**Comment by laurenmurphyx64** [comment]
@carlescufi @tejlmand I'm sorry to ping - I figure the two of you are probably doing a lot of code review on top of other work at the moment. But since this PR is a feature, it looks like it will have to be merged before Feature Freeze on Friday, May 7th. If you have time this week or next to give a review, it would be much appreciated.

**Review by carlescufi — APPROVED** [review]
I have no more objections myself, but I would like @tejlmand to review before we merge

File: samples/modules/tensorflow/hello_world/CMakeLists.txt
Comment (tejlmand) [inline samples/modules/tensorflow/hello_world/CMakeLists.txt:7]: Please make an entry in:
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/compiler/compiler_flags_template.cmake

and minimum for GCC:
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/compiler/gcc/compiler_flags.cmake
Clang supports the same flag and inherits its base from the gcc flags.

@evgeniy-paltsev does MWDT support the same flag ?

File: west.yml
Comment (laurenmurphyx64) [reply west.yml]: @hakehuang @nashif Moved west.yml reference to the new [Zephyr-hosted TensorFlow repo](https://github.com/zephyrproject-rtos/tensorflow).

**Review by tejlmand — COMMENTED** [review]
Build files and placement looks good.
The `-fno-threadsafe-statics` flag should go to out compiler flags, so that it will be easier to support multiple toolchains.

**Review by nashif — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

File: samples/modules/tensorflow/hello_world/CMakeLists.txt
Comment (evgeniy-paltsev) [reply samples/modules/tensorflow/hello_world/CMakeLists.txt:7]: Yep, ARC MWDT supports the `-fno-threadsafe-statics` flag

**Pull request merged**
